File: tests/test_parse/data/sample.erl
-module(user_manager).
-export([create_user/2, process_user/1, handle_users/1, start/0]).

% Record definition
-record(user, {name, age}).

% Type specifications
-type user() :: #user{}.
-type result() :: {ok, user()} | {error, term()}.

% Function with pattern matching and guards
-spec create_user(string(), integer()) -> user().
create_user(Name, Age) when is_list(Name), is_integer(Age), Age >= 0 ->
    #user{name = Name, age = Age}.

% Function with multiple clauses
-spec process_user(user()) -> result().
process_user(#user{name = Name, age = Age}) when Age >= 18 ->
    io:format("Processing adult user: ~s~n", [Name]),
    {ok, #user{name = Name, age = Age}};
process_user(#user{name = Name}) ->
    io:format("Processing minor user: ~s~n", [Name]),
    {error, underage}.

% Function using list comprehension
-spec handle_users([user()]) -> [result()].
handle_users(Users) ->
    [process_user(User) || User <- Users].

% Function using case expression
validate_age(Age) ->
    case Age of
        A when is_integer(A), A >= 0 -> {ok, A};
        _ -> {error, invalid_age}
    end.

% Private helper function (not exported)
format_user(#user{name = Name, age = Age}) ->
    lists:flatten(io_lib:format("~s (~p years old)", [Name, Age])).

% Function with try-catch
safe_process_user(User) ->
    try process_user(User) of
        Result -> Result
    catch
        error:Error -> {error, Error};
        _:_ -> {error, unknown_error}
    end.

% Function using higher-order functions
process_names(Names) ->
    lists:map(fun(Name) -> string:uppercase(Name) end, Names).

% Function returning fun
create_greeter(Greeting) ->
    fun(Name) -> Greeting ++ ", " ++ Name ++ "!" end.

% Function with message passing
start_user_processor() ->
    spawn(fun() -> user_processor_loop() end).

user_processor_loop() ->
    receive
        {process, User} ->
            process_user(User),
            user_processor_loop();
        stop ->
            ok
    end.

% Main function to demonstrate usage
start() ->
    % Create some users
    User1 = create_user("John", 25),
    User2 = create_user("Alice", 17),
    
    % Process individual users
    process_user(User1),
    process_user(User2),
    
    % Process list of users
    Users = [User1, User2],
    Results = handle_users(Users),
    io:format("Processing results: ~p~n", [Results]),
    
    % Try higher-order functions
    Names = ["bob", "alice", "charlie"],
    UpperNames = process_names(Names),
    io:format("Uppercase names: ~p~n", [UpperNames]),
    
    % Try function generator
    Greeter = create_greeter("Hello"),
    Greeting = Greeter("Dave"),
    io:format("~s~n", [Greeting]),
    
    % Try message passing
    Pid = start_user_processor(),
    Pid ! {process, User1},
    Pid ! stop,
    
    ok. 

File: tests/test_parse/data/sample.ex
defmodule UserManager do
  # Module attribute
  @default_age 18

  # Public function with pattern matching
  def process_user(%{name: name, age: age} = user) when age >= @default_age do
    IO.puts("Processing adult user: #{name}")
    {:ok, user}
  end

  def process_user(%{name: name}) do
    IO.puts("Processing minor user: #{name}")
    {:error, :underage}
  end

  # Private function
  defp validate_age(age) when is_integer(age) and age > 0 do
    {:ok, age}
  end

  defp validate_age(_), do: {:error, :invalid_age}

  # Function with default arguments
  def create_user(name, age \\ @default_age) do
    case validate_age(age) do
      {:ok, valid_age} -> %{name: name, age: valid_age}
      error -> error
    end
  end

  # Anonymous function usage
  def process_list(list) do
    Enum.map(list, fn item ->
      String.upcase(item)
    end)
  end

  # Function returning anonymous function
  def create_greeter(greeting) do
    fn name -> "#{greeting}, #{name}!" end
  end

  # Function with guard clauses
  def check_age(age) when is_integer(age) and age >= @default_age do
    :adult
  end

  def check_age(age) when is_integer(age) do
    :minor
  end

  def check_age(_) do
    :invalid
  end

  # Pipe operator example function
  def process_name(name) do
    name
    |> String.trim()
    |> String.capitalize()
    |> String.replace(" ", "_")
  end

  # Function with multiple clauses and pattern matching
  def handle_result({:ok, value}), do: "Success: #{value}"
  def handle_result({:error, reason}), do: "Error: #{reason}"
  def handle_result(_), do: "Unknown result"

  # Function using comprehension
  def process_users(users) do
    for user <- users,
        Map.has_key?(user, :name),
        do: process_user(user)
  end
end

# Protocol implementation example
defprotocol Formatter do
  def format(value)
end

defimpl Formatter, for: Map do
  def format(map) do
    map
    |> Map.to_list()
    |> Enum.map(fn {k, v} -> "#{k}: #{v}" end)
    |> Enum.join(", ")
  end
end

# Example usage
defmodule Example do
  def run do
    user = UserManager.create_user("John", 25)
    UserManager.process_user(user)

    greeter = UserManager.create_greeter("Hello")
    IO.puts(greeter.("Alice"))

    users = [
      %{name: "Bob", age: 30},
      %{name: "Alice", age: 25},
      %{name: "Charlie", age: 15}
    ]
    UserManager.process_users(users)
  end
end 